source/extensions/filters/http/match_delegate/config.cc

```diff
@@ -121,7 +121,7 @@ void DelegatingStreamFilter::FilterMatchState::evaluateMatchTree(
 
   if (match_tree_evaluated_ && match_result.result_) {
     const auto result = match_result.result_();
```

Comment:
@yanjunxiang-google , I think the sampling is actually performed at `match_result.result_()` as callback, right?

My comment above are :
(1) Since the sampling is already performed in match_delegate filter, It seems to be more natural to have the sampling config in its own filter config?
(2) It is more generic as other `TypedExtensionConfig` besides composite action can also make use of sampling feature under match_delegate.

Reply:
cc @htuch  and @rshriram for second opinion

Reply:
we can also interpret the current approach has done sampling inside the action callbacks, and match_delegate filter  is not aware of sampling or whatsoever, and it does not have to know.

Reply:
Also, I feel sampling fits well for the composite filter logic:  if sampling returns positive, then take the action, otherwise, skip.  So, I guess the existing solution resolves both https://github.com/envoyproxy/envoy/issues/30147 and https://github.com/envoyproxy/envoy/issues/30006.

Another thing is that adding the sampling under the wrapper of xDS matcher requires to add the sample_percent API in both ExtensionWithMatcher and ExtensionWithMatcherPerRouter.  From API perspective, not sure whether this is clean or not.

Reply:
But the current approach is not generic because it only limits to composite action callback. If other actions need sampling, you will need to either duplicate sampling or force them to use composite action.

Also, as change here shows, match_delegate must invoke the callback and check the result to make sampling work. So it is not transparent to match_delegate .

Reply:
Yeah, I think ask customer to use composite action probably is okay for the existing issues https://github.com/envoyproxy/envoy/issues/30147 and https://github.com/envoyproxy/envoy/issues/30006.  As stated above, composite filter has the characteristics either executing the action or skipping is a good fit for the sampling behavior. 

For the change here, adding a result nullptr check is intuitive to me before call result->typeUrl(). We should add this nullptr check regardless whether the underline filter callback has sampling or not.

There is another issue if putting the sampling under the ExtensionWithMatcher but not in composite action:  Let's say you have a matcher list with three matchers, and one matcher to ext_proc, which you want to do sampling with 10% ratio; Another matcher to RBAC filter, which you want to do sampling with 100% ratio; The third matcher with access_logging, you want to do sampling with 1%.  Putting sampling under ExtensionWithMatcher wont' support such kind of flexibility as you only have one ratio for all underline matchers. But putting it under action can.

Reply:
I think `ExtensionWithMatcher` can still support the example above (which I agree it's a legit use case) with a map of sampling ratios. Action name will be the key to look up ratio for matched action filter and sampling can be performed accordingly. This will introduce certain implementation complexity to match_delegate though. 
With that said, I am ok with current approach to unblock our existing needs. In the long term, we probably should think about xDS matcher approach to make it ultimately flexible.